components/datetime/src/provider.rs

```diff
@@ -0,0 +1,200 @@
+    fn get_symbol_for_weekday(
+        &self,
+        weekday: fields::Weekday,
+        length: fields::FieldLength,
+        day: usize,
+    ) -> &Cow<str>;
+    fn get_symbol_for_day_period(
+        &self,
+        day_period: fields::DayPeriod,
+        length: fields::FieldLength,
+        hour: usize,
+    ) -> &Cow<str>;
+}
+
+impl DateTimeDates for structs::dates::gregory::DatesV1 {
+    fn get_pattern_for_options(&self, options: &DateTimeFormatOptions) -> Result<Option<Pattern>> {
+        match options {
+            DateTimeFormatOptions::Style(bag) => self.get_pattern_for_style_bag(bag),
+            DateTimeFormatOptions::Components(_) => unimplemented!(),
+        }
+    }
+
+    fn get_pattern_for_style_bag(&self, style: &style::Bag) -> Result<Option<Pattern>> {
+        match (style.date, style.time) {
+            (None, None) => Ok(None),
+            (None, Some(time_style)) => self.get_pattern_for_time_style(time_style).map(Some),
+            (Some(date_style), None) => self.get_pattern_for_date_style(date_style).map(Some),
+            (Some(date_style), Some(time_style)) => {
+                let time = self.get_pattern_for_time_style(time_style)?;
+                let date = self.get_pattern_for_date_style(date_style)?;
+
+                self.get_pattern_for_date_time_style(date_style, date, time)
+                    .map(Some)
+            }
+        }
+    }
+
+    fn get_pattern_for_date_style(&self, style: style::Date) -> Result<Pattern> {
+        let date = &self.patterns.date;
+        let s = match style {
+            style::Date::Full => &date.full,
+            style::Date::Long => &date.long,
+            style::Date::Medium => &date.medium,
+            style::Date::Short => &date.short,
+        };
+        Ok(Pattern::from_bytes(s.as_bytes())?)
+    }
+
+    fn get_pattern_for_date_time_style(
+        &self,
+        style: style::Date,
+        date: Pattern,
+        time: Pattern,
+    ) -> Result<Pattern> {
+        let date_time = &self.patterns.date_time;
+        let s = match style {
+            style::Date::Full => &date_time.full,
+            style::Date::Long => &date_time.long,
+            style::Date::Medium => &date_time.medium,
+            style::Date::Short => &date_time.short,
+        };
+        Ok(Pattern::from_bytes_combination(s.as_bytes(), date, time)?)
+    }
+
+    fn get_pattern_for_time_style(&self, style: style::Time) -> Result<Pattern> {
+        let time = &self.patterns.time;
+        let s = match style {
+            style::Time::Full => &time.full,
+            style::Time::Long => &time.long,
+            style::Time::Medium => &time.medium,
+            style::Time::Short => &time.short,
+        };
+        Ok(Pattern::from_bytes(s.as_bytes())?)
+    }
+
+    fn get_symbol_for_weekday(
+        &self,
+        weekday: fields::Weekday,
+        length: fields::FieldLength,
+        day: usize,
+    ) -> &Cow<str> {
+        debug_assert!(day < 7);
```

Comment:
It's not an assertion.  Assertions should be assertions.  Since we allow the user to pass a `usize` as their weekday in DummyDataProvider, we should fail with an error, not crash.

Reply:
If you want to make it an assertion in this module, that's fine, but you should perform the validation sooner.

As I suggested in #158, consider solving this problem by changing the `usize` in DummyFields to a type that is only capable of representing numbers 1-7 (or 0-6).

Reply:
I added `Error::Data(String)` with "Month must be 0-11" and "Day of the week must be 0-6" which will be returned from the formatted before the value is passed here. I kept the debug assertions just in case.

If you don't like this solution, please, write what you want to see.

Reply:
Damn, github is failing me. It shows some of your comments some of the time. I just saw the last one which I didn't see before. I apologize for my last comment. I'll implement that.